cli/pcluster/config/param_types.py

```diff
@@ -0,0 +1,1202 @@
+# The following classes represent the Param of the standard types
+# like String, Int, Float, Bool and Json
+# and how to convert them from/to CFN/file.
+
+
+class Param(object):
+    """Class to manage simple string configuration parameters."""
+
+    def __init__(
+        self,
+        section_key,
+        section_label,
+        param_key,
+        param_map,
+        pcluster_config,
+        cfn_value=None,
+        config_parser=None,
+        cfn_params=None,
+    ):
+        self.section_key = section_key
+        self.section_label = section_label
+        self.key = param_key
+        self.map = param_map
+        self.pcluster_config = pcluster_config
+
+        # initialize param value
+        if cfn_params or cfn_value:
+            self._init_from_cfn(cfn_params=cfn_params, cfn_value=cfn_value)
+        elif config_parser:
+            try:
+                self._init_from_file(config_parser)
+            except NoOptionError:
+                self._init_from_map()
+            except NoSectionError:
+                section_name = _get_file_section_name(self.section_key, self.section_label)
+                error("Section '[{0}]' not found in the config file.".format(section_name))
+        else:
+            self._init_from_map()
+
+    def get_value_from_string(self, string_value):
+        """Return internal representation starting from CFN/user-input value."""
+        param_value = self.get_default_value()
+
+        if isinstance(string_value, str):
+            string_value = string_value.strip()
+
+        if string_value and string_value != "NONE":
+            param_value = string_value
+
+        return param_value
+
+    def _init_from_file(self, config_parser):
+        """
+        Initialize param_value from config_parser.
+
+        :param config_parser: the configparser object from which get the parameter
+        :raise NoOptionError, NoSectionError if unable to get the param from config_parser
+        """
+        section_name = _get_file_section_name(self.section_key, self.section_label)
+        self.value = config_parser.get(section_name, self.key)
+        self._check_allowed_values()
+
+    def _init_from_cfn(self, cfn_params=None, cfn_value=None):
+        """
+        Initialize parameter value by parsing CFN input parameters or from a given value coming from CFN.
+
+        :param cfn_params: the list of all the CFN parameters, it is used if "cfn" attribute is specified in the map
+        :param cfn_value: a value coming from a comma separated CFN param
+        """
+        cfn_converter = self.map.get("cfn", None)
+        if cfn_converter and cfn_params:
+            cfn_value = get_cfn_param(cfn_params, cfn_converter) if cfn_converter else "NONE"
+            self.value = self.get_value_from_string(cfn_value)
+        elif cfn_value:
+            self.value = self.get_value_from_string(cfn_value)
+        else:
+            self._init_from_map()
+
+    def _init_from_map(self):
```

Comment:
this seems to be a simple wrapper around get_default_value. Can we use get_default_value directly or at least pick a more meaningful name?